File: exercism/javascript/complete/anagram/anagram.js
// refactor to use hash map solution

class Anagram {
  constructor(word) {
    this.word = word;
  }

  matches(arr) {
    let anagrams = [];
    let index = 0;
    if (this.word.match(/[A-Z]/g) != null && this.word.match(/[A-Z]/g).length > 1) return anagrams;
    let match = this.word.toLowerCase().split('').sort();
    while (index < arr.length) {
      let text = arr[index].toLowerCase().split('').sort();
      let wrong = 0;
      if (text.length == match.length) {
        for (let i=0; i<match.length; i++) {
          if (text[i] != match[i]) {
            wrong++;
          }
        }
      } else {
        wrong = text.length - match.length;
      }
      if (wrong == 0) {
        anagrams.push(arr[index]);
      }
      index++;
    }
    return anagrams;
  }
}

export default Anagram;

File: exercism/ruby/complete/anagram/anagram.rb
class Anagram

  def initialize(word)
    letter_count = Hash.new(0)
    @word = word.downcase
    @word.chars.each{ |letter| letter_count[letter] += 1 }
    @letter_count = letter_count.sort_by{ |k,v| k }
  end

  def match(array) #return word in array that are anagrams of @letter_count
    output = []
    array.each do |word|
      hash = Hash.new(0)
      word.chars.each { |letter| hash[letter.downcase] += 1 }
      output << word if @letter_count == hash.sort_by{ |k,v| k } && word.downcase != @word
    end
    output
  end
end
